cvat/apps/engine/user_deletion.py

```diff
@@ -0,0 +1,122 @@
+# Copyright (C) CVAT.ai Corporation
+#
+# SPDX-License-Identifier: MIT
+
+from abc import ABC, ABCMeta, abstractmethod
+from dataclasses import dataclass
+from typing import TypedDict
+
+from django.db import transaction
+from django.db.models import Count, Max, Q, QuerySet
+from django.db.models.functions import Coalesce
+from rest_framework.exceptions import ValidationError
+
+from cvat.apps.engine.models import CloudStorage, Project, Task, User
+from cvat.apps.organizations.models import Organization
+
+_USER_DELETION_VALIDATORS = []
+
+
+class AutoRegisterValidatorMeta(ABCMeta):
+    def __new__(mcs, name, bases, namespace):
+        cls = super().__new__(mcs, name, bases, namespace)
+        if not getattr(cls, "__abstractmethods__", None):
+            _USER_DELETION_VALIDATORS.append(cls)
+        return cls
+
+
+class UserDeletionValidator(ABC, metaclass=AutoRegisterValidatorMeta):
+    """
+    Base class for user deletion validators.
+    To add a new validator, inherit from this class and implement the `validate` method.
+    The implemented class will be automatically registered and used during user deletion.
+    """
+
+    @abstractmethod
+    def validate(self, user: User) -> None:
+        """
+        Perform validation before user deletion.
+
+        If the user cannot be deleted, this method must raise
+        rest_framework.exceptions.ValidationError with a description of the reason.
+        """
+
+
+class NonEmptyOrgsValidator(UserDeletionValidator):
+    def validate(self, user: User) -> None:
+        max_members = (
+            Organization.objects.filter(owner=user)
+            .annotate(members_count=Count("members"))
+            .aggregate(max_members=Coalesce(Max("members_count"), 0))["max_members"]
+        )
+
+        if max_members > 1:
+            raise ValidationError(
+                "Cannot delete a user who is the owner of an organization with multiple members."
```

Comment:
Consider adding the org id in the error message, it can help if there are several owned orgs.